File: 백준/Gold/1339. 단어 수학/단어 수학.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        arr = new String[N];
        for (int i = 0; i < N; i++) {
        	arr[i] = br.readLine();
        	addWeight(arr[i]);
        }
        
        
        for(int i = 0; i < 26; i++) {
        	if(weights[i] != 0) {
        		char ch = (char)(i+'A');
        		pq.add(new Pair(ch, weights[i]));
        	}
        }
        
        int num = 9;
        
        while(!pq.isEmpty()) {
        	Pair cur = pq.poll();
        	weights[toInt(cur.ch)] = num;
        	num--;
        }
        
        long ans = 0;
        
        for(int i = 0; i < N; i++) {
        	ans += convert(arr[i]);
        }
        System.out.println(ans);
    }
    
    static long convert(String str) {
    	long ret = 0;
    	
    	for(int i = 0; i < str.length();i++) {
    		char ch = str.charAt(i);
    		
    		ret = ret * 10 + weights[toInt(ch)];
    	}
    	
    	return ret;
    }
    
    static PriorityQueue<Pair> pq = new PriorityQueue<>();
    
    static class Pair implements Comparable<Pair>{
    	char ch;
    	long value;
    	
    	public Pair(char ch,long value) {
    		this.ch = ch;
    		this.value = value;
    	}
    	
    	public int compareTo(Pair o) {
    		return Long.compare(o.value, value);
    	}
    }
    
    static long weights[] = new long[26];
    
    static int toInt(char ch) {
    	return ch - 'A';
    }
    
    static void addWeight(String str) {
    	int len = str.length();
    	for(int i = 0; i < len;i++) {
    		char ch = str.charAt(i);
    		int pos = len - i;
    		weights[toInt(ch)] += (long)Math.pow(10, pos);
    	}
    }
    
    static int N;
    static String arr[];
}

File: 백준/Gold/1339. 단어 수학/단어 수학.py
from collections import defaultdict
from heapq import heappop, heappush

def addWeight(string):
    length = len(string)

    for i in range(length):
        ch = string[i]
        pos = length - i
        # print(10 ** pos)
        weights[ch] += 10 ** pos


weights = defaultdict(int)
class Pair:
    def __init__(self, ch, value):
        self.ch = ch
        self.value = value

    def __lt__(self, o):
        return self.value > o.value

    def __repr__(self):
        return f"ch: {self.ch}, value: {self.value}"

def convert(string):
    ret = 0
    for i in range(len(string)):
        ch = string[i]
        ret = ret * 10 + weights[ch]

    return ret

if __name__ == '__main__':
    N = int(input())
    arr = []
    for i in range(N):
        arr.append(input())
        addWeight(arr[i])

    pq = []
    for key, value in weights.items():

        heappush(pq, Pair(key,value))

    num = 9
    while pq:
        cur = heappop(pq)
        weights[cur.ch] = num
        num -= 1

    

    ans = 0
    for i in range(N):
        ans += convert(arr[i])

    print(ans)